pkg/innerring/node.go

```diff
@@ -0,0 +1,354 @@
+// contract must be configured by 'contracts.neofs' path. In this case if
+// mainChainStatus.notaryServiceEnabled is true, then address of the Processing
+// contract must be configured by 'contracts.processing' path.
+func newNode(cfg *viper.Viper, nodeAcc *wallet.Account, log *logger.Logger, chErr chan<- error, mainChain mainChainStatus) (*node, error) {
+	var n node
+	var err error
+
+	n.sidechain, err = blockchain.New(cfg.ConfigFileUsed(), nodeAcc, log.Named("Morph"), chErr)
+	if err != nil {
+		return nil, fmt.Errorf("init blockchain module: %w", err)
+	}
+
+	const contractsConfigPrefix = "contracts."
+	mConfigContracts := make(map[contractName]string)
+
+	if mainChain.separated {
+		mConfigContracts[contractNeoFS] = contractsConfigPrefix + "neofs"
+
+		if mainChain.notaryServiceEnabled {
+			mConfigContracts[contractProcessing] = contractsConfigPrefix + "processing"
+		}
+	}
+
+	n.contracts, err = newContracts(cfg, mConfigContracts)
+	if err != nil {
+		return nil, fmt.Errorf("init blockchain contracts: %w", err)
+	}
+
+	n.storageEmissionAmount = cfg.GetUint64("emit.storage.amount")
+	n.acc = nodeAcc
+
+	n.processors.alphabet = alphabet.New(log, &n)
+	n.processors.balance = balance.New(log, &n)
+	n.processors.container = container.New(log, &n)
+
+	n.mainChain.enabled = mainChain.separated
+	n.log = log
+
+	return &n, nil
+}
+
+// run starts underlying blockchain service and starts background process which
+// listens to incoming events of the underlying blockchain and handles the
+// NeoFS-specified ones. The listen process is aborted by the specified channel.
+// Returns any error encountered which prevented the node to be started.
+// If run failed, the node should no longer be used. After node has
+// been successfully run, all internal failures are written to the channel
+// provided to newNode.
+//
+// The run method should not be called more than once.
+//
+// Use stop method to stop the node.
+func (x *node) run(chDone <-chan struct{}) error {
+	err := x.blockChain.run(chDone)
+	if err != nil {
+		return fmt.Errorf("run node's blockchain basement: %w", err)
+	}
+
+	go x.listenBlockchain(chDone)
+
+	return nil
+}
+
+// stop stops the running node and frees all its internal resources.
+//
+// The stop method should not be called twice and before successful run.
+func (x *node) stop() {
+	x.blockChain.stop()
+}
+
+// IsAlphabet implements common.NodeStatus via checking if local Inner Ring
+// node's account belongs to someone from the NeoFS Sidechain committee.
+func (x *node) IsAlphabet() (bool, error) {
+	letter, err := x.alphabetLetterForAccount(x.acc)
+	if err != nil {
+		return false, err
+	}
+
+	return letter >= 0, nil
+}
+
+// EmitSidechainGAS implements alphabet.LocalNode via triggering production of
+// the sidechain GAS and its distribution among Inner Ring nodes and Proxy
+// contract by calling 'emit' method of the Alphabet contract associated with
+// the local Inner Ring node.
+func (x *node) EmitSidechainGAS() error {
```

Comment:
I try to make processors to perform logical handling, not a thing where events go irrevocably. Direct use of actors complicates maintainability, because the control over events' postprocessing should be handled by the node, not by a single processor IMO.